src/lib/table/table-data-source.ts

```diff
@@ -0,0 +1,209 @@
+   * table's rendered data.
+   *
+   * Note that the data source uses the paginator's properties to calculate which page of data
+   * should be displayed. If the paginator receives its properties as template inputs,
+   * e.g. `[pageLength]=100` or `[pageIndex]=1`, then be sure that the paginator's view has been
+   * initialized before assigning it to this data source.
+   */
+  set paginator(paginator: MdPaginator|null) {
+    this._paginator = paginator;
+    this._updateChangeSubscription();
+  }
+  get paginator(): MdPaginator|null { return this._paginator; }
+  private _paginator: MdPaginator|null;
+
+  /**
+   * Data accessor function that is used for accessing data properties for sorting.
+   * This default function assumes that the sort header IDs (which defaults to the column name)
+   * matches the data's properties (e.g. column Xyz represents data['Xyz']).
+   * May be set to a custom function for different behavior.
+   * @param data Data object that is being accessed.
+   * @param sortHeaderId The name of the column that represents the data.
+   */
+  sortingDataAccessor = (data: T, sortHeaderId: string): string|number => {
+    const value: number|string = data[sortHeaderId];
+    return isNaN(+value) ? value : +value;
+  }
+
+  /**
+   * Transforms data objects into a filter term that will be used to check against the filter if
+   * a filter is set. By default, the function will iterate over the values of the data object
+   * and convert them to a lowercase string.
+   * @param data Data object to convert to a string that checked for containing the filter term.
+   */
+  filterTermAccessor: ((data: T) => string) = (data: T): string => {
+    const accumulator = (currentTerm, key) => currentTerm + data[key];
+    return Object.keys(data).reduce(accumulator, '').toLowerCase();
+  }
+
+  constructor(initialData: T[] = []) {
+    this._data = new BehaviorSubject<T[]>(initialData);
+    this._updateChangeSubscription();
+  }
+
+  /**
+   * Subscribe to changes that should trigger an update to the table's rendered rows. When the
+   * changes occur, process the current state of the filter, sort, and pagination along with
+   * the provided base data and send it to the table for rendering.
+   */
+  _updateChangeSubscription() {
+    // Sorting and/or pagination should be watched if MdSort and/or MdPaginator are provided.
+    // Otherwise, use an empty observable stream to take their place.
+    const sortChange = this._sort ? this._sort.mdSortChange : observableOf();
+    const pageChange = this._paginator ? this._paginator.page : observableOf();
+
+    if (this._renderChangesSubscription) { this._renderChangesSubscription.unsubscribe(); }
+    this._renderChangesSubscription = RxChain.from(this._data)
+        // Watch for base data or filter changes to provide a filtered set of data.
+        .call(combineLatest, this._filter)
+        .call(map, ([data]) => this._filterData(data.slice()))
```

Comment:
Why do you clone the array on every call?

Reply:
From `_updateChangeSubscription`'s perspective, those calls may change the data in some way (filter, reorder, or splice). When an event triggers, e.g. sort order changed, it wants to make sure `data` is preserved so it sends a copy to be processed. If sort order changes again, it knows that data array provided on the stream has been protected from mutation.

The exception here is that `filterData` doesn't actually mutate the data array, but it felt unsymmetrical to "know" that it can be passed the original array safely.

Reply:
I would think that the cloning logic more belongs inside the functions that here in the chain

Reply:
SGTM - not too opinionated on it. Moved cloning to the functions themselves